codex/contracts/provider.nim

```diff
@@ -4,34 +4,29 @@ import pkg/questionable
 
 import ../logutils
 
-from ../errors import CodexError
 from ../clock import SecondsSince1970
 
 logScope:
   topics = "marketplace onchain provider"
 
-type CodexProviderError* = object of CodexError
-
-proc raiseCodexProviderError(message: string) {.raises: [CodexProviderError].} =
-  raise newException(CodexProviderError, message)
+proc raiseProviderError(message: string) {.raises: [ProviderError].} =
+  raise newException(ProviderError, message)
```

Comment:
We cannot use:

```nim
without latestBlock =? await provider.getBlock(blockTag), error:
    raiseCodexProviderError("Could not get latest block", error)
```

because (as you yourself pointed out earlier) `getBlock` returns an `Option` and  not a `Result`. Right?

So if there was an original `ProviderError` caused by `getBlock` it will naturally bubble-up so, we do not have to do anything about it I suppose. But then `getBlock`, which is defined as this:

```nim
method getBlock*(
  provider: JsonRpcProvider,
  tag: BlockTag): Future[?Block] {.async: (raises:[ProviderError]).} =

  convertError:
    let client = await provider.client
    return await client.eth_getBlockByNumber(tag, false)
```

can still return `None` because `eth_getBlockByNumber` is defined like this:

```nim
proc eth_getBlockByNumber(blockTag: BlockTag, includeTransactions: bool): ?Block
```

This is why I am raising `ProviderError` when this happen, but we have no data about underlying exception that caused `eth_getBlockByNumber` to return `None`.

So I think in this case everything can stays as it is, right?

Reply:
Good point, sorry I missed that `getBlock` returns an `Option` 👍 

> we have no data about underlying exception that caused eth_getBlockByNumber to return None
I don't think it's an error really, just that the ethereum client can return null if the block doesn't exist.

What you have looks good 👍